sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs

```diff
@@ -0,0 +1,101 @@
+    /// </summary>
+    internal class StorageBearerTokenChallengeAuthorizationPolicy : BearerTokenAuthenticationPolicy
+    {
+        private readonly string[] _scopes;
+        private volatile string tenantId;
+        private readonly bool _enableTenantDiscovery;
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="credential"></param>
+        /// <param name="scope"></param>
+        /// <param name="enableTenantDiscovery"> </param>
+        public StorageBearerTokenChallengeAuthorizationPolicy(TokenCredential credential, string scope, bool enableTenantDiscovery) : base(credential, scope)
+        {
+            Argument.AssertNotNullOrEmpty(scope, nameof(scope));
+            _scopes = new[] { scope };
+            _enableTenantDiscovery = enableTenantDiscovery;
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="credential"></param>
+        /// <param name="scopes"></param>
+        /// <param name="enableTenantDiscovery"> </param>
+        public StorageBearerTokenChallengeAuthorizationPolicy(TokenCredential credential, IEnumerable<string> scopes, bool enableTenantDiscovery) : base(
+            credential,
+            scopes)
+        {
+            Argument.AssertNotNull(scopes, nameof(scopes));
+            _scopes = scopes.ToArray();
+            _enableTenantDiscovery = enableTenantDiscovery;
+        }
+
+        /// <inheritdoc />
+        protected override void AuthorizeRequest(HttpMessage message)
+        {
+            if (tenantId != null || !_enableTenantDiscovery)
+            {
+                base.AuthorizeRequest(message);
+            }
+        }
+
+        /// <inheritdoc />
+        protected override ValueTask AuthorizeRequestAsync(HttpMessage message)
+        {
+            if (tenantId != null || !_enableTenantDiscovery)
+            {
+                return base.AuthorizeRequestAsync(message);
+            }
+            return default;
+        }
+
+        /// <inheritdoc />
+        protected override bool AuthorizeRequestOnChallenge(HttpMessage message) => AuthorizeRequestOnChallengeInternalAsync(message, false).EnsureCompleted();
+
+        /// <inheritdoc />
+        protected override ValueTask<bool> AuthorizeRequestOnChallengeAsync(HttpMessage message) => AuthorizeRequestOnChallengeInternalAsync(message, true);
+
+        private async ValueTask<bool> AuthorizeRequestOnChallengeInternalAsync(HttpMessage message, bool async)
+        {
+            try
+            {
+                var authUri = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "authorization_uri");
+
+                // tenantId should be the guid as seen in this example: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/authorize
+                tenantId = new Uri(authUri).Segments[1].Trim('/');
```

Comment:
why do we cache tenantId if we always parse it out?

Reply:
We only parse it on the first request. Is there a scenario where it would change for a given client after the initial challenge?

Reply:
Is this how flow looks like assuming that tenant discovery is enabled?
1. first request comes, AutorizeRequest is no-op which leads to 401. (because tenantId is null)
2. challenge logic (this function) kicks is, presumably producing working token.
3. tenantId as well as token is cached and AuthorizeRequest isn't no-op, so subsequent request just reuse cached token.

Questions that come to my mind:
- Is it right to make AuthorizeRequest no-op for first request? What if user provided TokenCredential that has correct tenant? Shouldn't this have a chance to work without invoking tenant discovery? Doesn't this lead to unnecessary extra request in that case?
- When cached token expires (assuming it's not refreshed for some reason) then, presumably, is this logic going to be executed again (assuming it got 401, not sure if with the challenge). If that scenario exist - should we leverage fact that we already parsed out the token instead of parsing it again?

Reply:
> Is this how flow looks like assuming that tenant discovery is enabled?
> 
> 1. first request comes, AutorizeRequest is no-op which leads to 401. (because tenantId is null)
> 2. challenge logic (this function) kicks is, presumably producing working token.
> 3. tenantId as well as token is cached and AuthorizeRequest isn't no-op, so subsequent request just reuse cached token.

Correct.

> Questions that come to my mind:
> 
> * Is it right to make AuthorizeRequest no-op for first request? What if user provided TokenCredential that has correct tenant? Shouldn't this have a chance to work without invoking tenant discovery? Doesn't this lead to unnecessary extra request in that case?

If the client thinks they know the tenant, they would not need to enable discovery at all.

> * When cached token expires (assuming it's not refreshed for some reason) then, presumably, is this logic going to be executed again (assuming it got 401, not sure if with the challenge). If that scenario exist - should we leverage fact that we already parsed out the token instead of parsing it again?

We could, but this is rare in high request volume scenarios because the token gets refreshed in the background and generally avoids having to handle a challenge again.

Reply:
>> If the client thinks they know the tenant, they would not need to enable discovery at all.
I wouldn't assume that. They might enable tenant discovery as over-protective measure even though they have right tenant id. Or they may want to use it as a fallback.
On large scale this might be an issue - somebody has to pay for network.

Reply:
This would cost one additional request (if the option was enabled) per client lifetime. Do you think that would be a significant concern for customers?

Reply:
That depends on how do they manage client lifecycle.
Users who keep centralized singleton of top level client and traverse hierarchy down (i.e. inherit pipeline) shouldn't feel that.
However, users who create clients out of the blue would. With introduction of shared static transport (as default) we kind of created a promise that creating new instances is lightweight.

I wonder how hard would be to change it - i.e. have Authorize operational on first request.
Would this just work? i.e. would we get challange if token is invalid OR does absense of token trigger the challange on the server side?
If former then why not? if later then I image we'd need extra call for users who actually seek discovery - in that case let's keep it as is.

Reply:
There are some scenarios where issuing the request with the wrong tenantId would fail at the `GetToken` call before we ever attempt to send the request to get a challenge.

Reply:
Gotcha. let's keep it as is then.